io.openems.edge.greenconsumptionadvisor.api/src/io/openems/edge/greenconsumptionadvisor/api/GreenConsumptionAdvisorImpl.java

```diff
@@ -0,0 +1,193 @@
+				GreenConsumptionAdvisor.ChannelId.values() //
+		);
+	}
+
+	@Activate
+	private void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+
+		// TODO validate zip code
+		this.urlString = URL_API_RECOMMONDATION + config.zip_code();
+		this.timeNextActualization = LocalDateTime.now();
+	}
+
+	@Override
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+	}
+
+	@Override
+	public void run() throws OpenemsNamedException {
+
+		if (this.timeNextActualization.isBefore(LocalDateTime.now())) {
+			this.updateDataBuffer();
+			this.timeNextActualization = LocalDateTime.now().truncatedTo(ChronoUnit.HOURS).plusHours(1);
+		}
+		this.evaluateDataBuffer();
+		this.writeRecommondationToChannel();
+	}
+
+	private void evaluateDataBuffer() {
+		if (this.dataBuffer == null) {
+			return;
+		}
+		if (this.currentInformation == null) {
+			try {
+				this.currentInformation = this.getGridEmissionInformationFromJson(this.dataBuffer.get(0));
+			} catch (OpenemsNamedException e) {
+				e.printStackTrace();
+			}
+		}
+		// check if newer information available, if yes actualize and delete first
+		// element of buffer
+		LocalDateTime lastFullHour = LocalDateTime.now().truncatedTo(ChronoUnit.HOURS);
+		int i;
+		try {
+			for (i = 0; i < this.dataBuffer.size() - 1; i++) {
+				GridEmissionInformation bufferElement = this.getGridEmissionInformationFromJson(this.dataBuffer.get(i));
+				if (bufferElement.getTimestamp().equals(lastFullHour)) {
+					this.currentInformation = bufferElement;
+					break;
+				}
+			}
+			// Check if no bufferElement was found, then reset all
+			LocalDateTime timeLastElement = this.getGridEmissionInformationFromJson(this.dataBuffer.get(i))
+					.getTimestamp();
+			if (this.dataBuffer.size() - 1 == i && this.currentInformation.getTimestamp().isAfter(timeLastElement)) {
+				this.dataBuffer = null;
+				this.currentInformation = null;
+			}
+		} catch (OpenemsNamedException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private GridEmissionInformation getGridEmissionInformationFromJson(JsonElement jsonElement)
+			throws OpenemsNamedException {
+		long timestamp = JsonUtils.getAsLong(jsonElement, "time");
+		Instant instant = Instant.ofEpochMilli(timestamp);
+		// Convert the Instant to LocalDateTime in the system's default time zone
+		LocalDateTime dateTime = LocalDateTime.ofInstant(instant, ZoneId.systemDefault());
+		int co2 = JsonUtils.getAsInt(jsonElement, "co2");
+		String advice = JsonUtils.getAsString(jsonElement, "advice");
+		ConsumptionAdvice consAdvice = ConsumptionAdvice.UNDEFINED;
+		switch (advice) {
+		case "green":
+			consAdvice = ConsumptionAdvice.GREEN;
+			break;
+		case "yellow":
+			consAdvice = ConsumptionAdvice.YELLOW;
+			break;
+		case "red":
+			consAdvice = ConsumptionAdvice.RED;
+			break;
+		}
+		return new GridEmissionInformation(dateTime, consAdvice, co2);
+	}
+
+	private void updateDataBuffer() {
```

Comment:
The current error handling logs exceptions but still continues execution. For example, if updateDataBuffer() encounters an error, it falls back to using stale data. While this prevents crashes, implementing a retry mechanism or a fallback to a cached version could improve reliability.

Reply:
Good point. Currently it uses data from the buffer in case of error and retries in the next hour. If that is not good enough I will add functionality to retry multiple times.

Reply:
So if it fails 4 Times in 4 hours one an hour it uses 4 hour old data? :)

Reply:
That would be right. But I changed it now so that it retries every 5 minutes if an error occures. Otherwise it waits for